sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamTest.scala

```diff
@@ -429,7 +429,25 @@ trait StreamTest extends QueryTest with SharedSQLContext with TimeLimits with Be
     val defaultCheckpointLocation =
       Utils.createTempDir(namePrefix = "streaming.metadata").getCanonicalPath
     try {
-      startedTest.foreach { action =>
+      val actionIterator = startedTest.iterator.buffered
+      while (actionIterator.hasNext) {
+        // Synchronize sequential addDataMemory actions.
+        val addDataMemoryActions = ArrayBuffer[AddDataMemory[_]]()
+        while (actionIterator.hasNext && actionIterator.head.isInstanceOf[AddDataMemory[_]]) {
+          addDataMemoryActions.append(actionIterator.next().asInstanceOf[AddDataMemory[_]])
+        }
+        if (addDataMemoryActions.nonEmpty) {
+          val synchronizeAll = addDataMemoryActions
```

Comment:
This is some magic-ish code. Can you add a bit more comments on how this compose thing works?